lib/notifications/display.dart

```diff
@@ -341,46 +338,25 @@ class NotificationDisplayManager {
 
   static String _personKey(Uri realmUri, int userId) => "$realmUri|$userId";
 
-  static void _onNotificationOpened(NotificationResponse response) async {
-    final payload = jsonDecode(response.payload!) as Map<String, dynamic>;
-    final data = MessageFcmMessage.fromJson(payload);
-    assert(debugLog('opened notif: message ${data.zulipMessageId}, content ${data.content}'));
-    _navigateForNotification(data);
-  }
-
-  static void _handleNotificationAppLaunch(NotificationResponse? response) async {
-    assert(response != null);
-    if (response == null) return; // TODO(log) seems like a bug in flutter_local_notifications if this can happen
-
-    final payload = jsonDecode(response.payload!) as Map<String, dynamic>;
-    final data = MessageFcmMessage.fromJson(payload);
-    assert(debugLog('launched from notif: message ${data.zulipMessageId}, content ${data.content}'));
-    _navigateForNotification(data);
-  }
-
-  static void _navigateForNotification(MessageFcmMessage data) async {
+  /// Navigates to the [MessageListPage] of the corresponding conversation
+  /// given the notification-open intent uri.
+  static Future<bool> navigateForNotification(Uri uri) async {
+    assert(debugLog('opened notif: uri: $uri'));
     NavigatorState navigator = await ZulipApp.navigator;
     final context = navigator.context;
     assert(context.mounted);
-    if (!context.mounted) return; // TODO(linter): this is impossible as there's no actual async gap, but the use_build_context_synchronously lint doesn't see that
+    if (!context.mounted) return false; // TODO(linter): this is impossible as there's no actual async gap, but the use_build_context_synchronously lint doesn't see that
 
     final globalStore = GlobalStoreWidget.of(context);
-    final account = globalStore.accounts.firstWhereOrNull((account) =>
-      account.realmUrl == data.realmUri && account.userId == data.userId);
-    if (account == null) return; // TODO(log)
+    final result = await _narrowFromNotificationIntentUri(globalStore: globalStore, uri: uri);
```

Comment:
Hmm — I think this gets a bit sticky for the UX in the case where Zulip wasn't already running and you open a notification. (Or where it was already running but hadn't loaded data for the account the notification is for.) The trouble is that this adds an `await` for getting the per-account store, before we do any navigation.

In main, `data` contains all the information we need for constructing a `Narrow`; so this `_navigateForNotification` method synchronously constructs a `Narrow` and immediately calls `navigator.push` for the appropriate `MessageListPage`, with no waiting. If the data isn't already loaded, the user may then see a loading page for a while… but at least they do see a loading page.

In the current version of this PR, I believe what happens in that situation is that we'll get to this line, and to the `await globalStore.perAccount(…)` line inside `_narrowFromNotificationIntentUri`, and then it'll sit there for however long it takes for the data to load — and during that time, there'll be no feedback to the user indicating that the notification did anything different from just launching the app.

(Well, I guess if the notification came from the first account in their list, then launching the app would have started that account's data loading and shown a corresponding loading page anyway. So in that case, which is the most common case because (presumably) most users have only one account, it'll work out fine. But if the notification is for an account that's later in the list, and if the app was already open or if the first account in the list is faster to load than the one the notification is for, the problem arises: we'll be showing a non-loading page that has nothing to do with the notification, with no visible indication that the notification had any effect, and then some time later the notification suddenly belatedly causes a navigation.)

---

It'd be possible to solve this problem head-on: basically we'd add a page that just takes the narrow URL as an argument, instead of a `Narrow`, and navigate there. Then that would show a loading page until the data arrives, at which point it'd replace itself with a normal MessageListPage. We'd need to take some care to avoid unnecessary page-transition animations, as well as an unnecessary flicker in the case where the data is already present.

We'll probably end up wanting some form of that head-on solution eventually, as part of handling Zulip links from outside the app, which is a post-launch feature.

But I think a simpler solution is that we can change the form of these `zulip://notification/…` URLs so that they directly encode the data we need: it's either a channel/stream ID and a topic, or a list of recipient IDs. Let's go with that solution for now, to keep down the complexity of this transition.